precompiles/xcm-utils/src/lib.rs

```diff
@@ -50,11 +60,35 @@ pub struct XcmUtilsPrecompile<Runtime, XcmConfig>(PhantomData<(Runtime, XcmConfi
 #[precompile_utils::precompile]
 impl<Runtime, XcmConfig> XcmUtilsPrecompile<Runtime, XcmConfig>
 where
-	Runtime: pallet_evm::Config + frame_system::Config,
+	Runtime: pallet_evm::Config + frame_system::Config + pallet_xcm::Config,
 	XcmOriginOf<XcmConfig>: OriginTrait,
 	XcmAccountIdOf<XcmConfig>: Into<H160>,
 	XcmConfig: xcm_executor::Config,
+	SystemCallOf<Runtime>: Dispatchable<PostInfo = PostDispatchInfo> + Decode + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::RuntimeCall as Dispatchable>::RuntimeOrigin:
+		From<Option<Runtime::AccountId>>,
+	<Runtime as frame_system::Config>::RuntimeCall: From<pallet_xcm::Call<Runtime>>,
 {
+	#[precompile::pre_check]
+	fn pre_check(handle: &mut impl PrecompileHandle) -> EvmResult {
+		// Only avoid calling execute from SC
+		if let Ok(selector) = handle.read_u32_selector() {
+			if XcmUtilsPrecompileCall::<Runtime, XcmConfig>::xcm_execute_selectors()
+				.contains(&selector)
+			{
+				handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+				let caller_code =
+					pallet_evm::Pallet::<Runtime>::account_codes(handle.context().caller);
+				// Check that caller is not a smart contract s.t. no code is inserted into
+				// pallet_evm::AccountCodes except if the caller is another precompile i.e. CallPermit
+				if !(caller_code.is_empty() || &caller_code == &[0x60, 0x00, 0x60, 0x00, 0xfd]) {
```

Comment:
Are we sure this wouldn't allow a SC to evade the check by doing a batch, one of the call opcode variants, etc.?

Reply:
this is what we do in other precompiles, so I guess not? if this is vulnerable, other precompiles are equally vulnerable